machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,174 @@
+	// by gathering the build inputs, validating them, and making sure we can do an update as specified.
+	MachineOSBuildPrepared BuildProgress = "Prepared"
+	// building indicates that the build has been kicked off with the specified image builder
+	MachineOSBuiling BuildProgress = "Building"
+	// failed indicates that during the build or preparation process, the build failed.
+	MachineOSBuildFailed BuildProgress = "Failed"
+	// interrupted indicates that the user stopped the build process by modifying part of the build config
+	MachineOSBuildInterrupted BuildProgress = "Interrupted"
+	// restarted indicates that this build has been either interrupted or failed, and now a new
+	// iteration of the process has begun.
+	MachineOSBuildRestarted BuildProgress = "Restarted"
+	// ready indicates that the build has completed and the image is ready to roll out.
+	MachineOSReady BuildProgress = "Ready"
+)
+
+// PriorMachineOSBuilds contains information about related builds
+type PriorMachineOSBuilds struct {
+	// name is the name of the build
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+	// buildFailure contains an optional message of why this build ended prematurely.
+	// +optional
+	BuildFailure string `json:"buildFailure"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the referenced object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret string `json:"baseImagePullSecret"`
```

Comment:
thought: I'm not opposed to keeping it around, but I'm not sure that we still need the `BaseImagePullSecret` field here anymore. I think the preferred way forward will be using the global pull secret instead since it should be able to pull the base OS image.

Reply:
Curious why the timeline comes into this decision? What would removing this have to do with the timeline?

Reply:
Ok so, perhaps there should be a spec field to allow the user to override, but a status field to show the actual secret that was used?

> don't think this logic of auto-retrieval will be done for 4.16

It sounds like you're thinking to ship this feature in parts, are there clear separations of the feature's within this that can and cannot be completed in time? Is there a way to make it clear what is, and is not supported for 4.16?